# Student Performance Indicator
### Life Cycle of Machine Learning Project

- Understanding the problem statement
- Data Collection
- Data Checks to Perform
- EDA
- Data Pre-processing
- Model Training
- Check the best model

# 1. Problem Statement

This project understand about how student performance for Test Score affected by Gender, Parental Level Education, Lunch and Test Preparation Score

# 2. Data Collection

Dataset Source is from Kaggle.com and the Data consist of 8 columns and 1000 rows

### 2.1 Import Data & Required Packages

Import Library

In [1]:
import pandas as pd
import numpy as np
import matplotlib as plt
import seaborn as sns
import warnings 
%matplotlib inline

warnings.filterwarnings('ignore')


Import CSV Data

In [2]:
df = pd.read_csv('data/stud.csv')

Show Top 5

In [3]:
df.head(5)

gender race_ethnicity parental_level_of_education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   

  test_preparation_course  math_score  reading_score  writing_score  
0                    none          72             72             74  
1               completed          69             90             88  
2                    none          90             95             93  
3                    none          47             57             44  
4                    none          76             78             75

Shape of Dataset

In [5]:
df.shape

(1000, 8)

### 2.2 Dataset Information

- gender : sex of students -> Male or Female
- race : race of students -> Group A,B,C,D,E
- parental level of education : parent's final education -> Bachelor's Degree, Some College, Master's Degree, Associate's Degree, School
- lunch : having lunch before test -> Standard or Free/Reduced
- test preparation course : Complete or not complete before test
- math score
- reading score
- writing score

# 3. Data Checks to Perform

- Check missing values
- Check duplicates
- Check data type
- Check the number of unique values of each column
- Check statistic of dataset
- Check various categories present in the different categorical column

### 3.1 Check Missing Values

In [6]:
df.isnull().sum()

gender                         0
race_ethnicity                 0
parental_level_of_education    0
lunch                          0
test_preparation_course        0
math_score                     0
reading_score                  0
writing_score                  0
dtype: int64

### 3.2 Check Duplicates

In [7]:
df.duplicated().sum()

np.int64(0)

### 3.3 Check Data Types

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   gender                       1000 non-null   object
 1   race_ethnicity               1000 non-null   object
 2   parental_level_of_education  1000 non-null   object
 3   lunch                        1000 non-null   object
 4   test_preparation_course      1000 non-null   object
 5   math_score                   1000 non-null   int64 
 6   reading_score                1000 non-null   int64 
 7   writing_score                1000 non-null   int64 
dtypes: int64(3), object(5)
memory usage: 62.6+ KB


### 3.4 Check Uniqueness Data

In [9]:
df.nunique()

gender                          2
race_ethnicity                  5
parental_level_of_education     6
lunch                           2
test_preparation_course         2
math_score                     81
reading_score                  72
writing_score                  77
dtype: int64

### 3.5 Check Statistic of Dataset

In [10]:
df.describe()

math_score  reading_score  writing_score
count  1000.00000    1000.000000    1000.000000
mean     66.08900      69.169000      68.054000
std      15.16308      14.600192      15.195657
min       0.00000      17.000000      10.000000
25%      57.00000      59.000000      57.750000
50%      66.00000      70.000000      69.000000
75%      77.00000      79.000000      79.000000
max     100.00000     100.000000     100.000000

### Insight
- All Means are very close to each other in range between 66 - 68.05
- All Standard Deviations are also close between 14.6 - 15.19
- 0 Minimum Score for Math and 10=> Score for Writing and Reading

### 3.6 EDA

In [19]:
df.head()

gender race_ethnicity parental_level_of_education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   

  test_preparation_course  math_score  reading_score  writing_score  
0                    none          72             72             74  
1               completed          69             90             88  
2                    none          90             95             93  
3                    none          47             57             44  
4                    none          76             78             75

In [21]:
print ("Categories in gender variable :", end=" ")
print (df['gender'].unique())

print ("Categories in race variable :", end=" ")
print (df['race_ethnicity'].unique())

print ("Categories in parental education level variable :", end=" ")
print (df['parental_level_of_education'].unique())

print ("Categories in lunch variable :", end=" ")
print (df['lunch'].unique())

print ("Categories in test preparation course variable :", end=" ")
print (df['test_preparation_course'].unique())

Categories in gender variable : ['female' 'male']
Categories in race variable : ['group B' 'group C' 'group A' 'group D' 'group E']
Categories in parental education level variable : ["bachelor's degree" 'some college' "master's degree" "associate's degree"
 'high school' 'some high school']
Categories in lunch variable : ['standard' 'free/reduced']
Categories in test preparation course variable : ['none' 'completed']


Categorical & Numerical Features

In [25]:
numeric_features = [feature for feature in df.columns if df[feature].dtype != 'O']
categorical_features = [feature for feature in df.columns if df[feature].dtype == 'O']

print (f"We have {len(numeric_features)} Numeric Features")
print (f"We have {len(categorical_features)} Categorical Features")

We have 3 Numeric Features
We have 5 Categorical Features


### 3.7 Adding Columns

In [27]:
df['total score'] = df['math_score'] + df['reading_score'] + df['writing_score']
df['average'] = df['total score'] / 3
df.head(5)

gender race_ethnicity parental_level_of_education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   

  test_preparation_course  math_score  reading_score  writing_score  \
0                    none          72             72             74   
1               completed          69             90             88   
2                    none          90             95             93   
3                    none          47             57             44   
4                    none          76             78             75   

   total score    average  
0          218  72.666667  
1          247  82.333333  
2          278  92.666667  
3          148  49.333333  
4          229  76.333333